src/confluent_kafka/src/Consumer.c

```diff
@@ -1356,17 +1479,67 @@ static void Consumer_rebalance_cb (rd_kafka_t *rk, rd_kafka_resp_err_t err,
 		}
 	}
 
+        if (!strcmp(rd_kafka_rebalance_protocol(rk), "COOPERATIVE")) {
+                if (self->u.Consumer.rebalance_assigned) {
+                        assign_errstr = "assign() must not be called during a "
+                                        "COOPERATIVE protocol rebalance";
+                } else if (self->u.Consumer.rebalance_incremental_unassigned &&
+                           err == RD_KAFKA_RESP_ERR__ASSIGN_PARTITIONS) {
+                        assign_errstr = "incremental_unassign() must not be "
+                                        "called during an ASSIGN rebalance ";
+                } else if (self->u.Consumer.rebalance_incremental_assigned &&
+                           err == RD_KAFKA_RESP_ERR__REVOKE_PARTITIONS) {
+                        assign_errstr = "incremental_assign() must not be "
+                                        "called during a REVOKE rebalance";
+                }
+        } else {
+                if (self->u.Consumer.rebalance_incremental_assigned ||
+                    self->u.Consumer.rebalance_incremental_unassigned) {
+                        assign_errstr = "incremental_(un)assign must not be "
+                                        "called during an EAGER protocol "
+                                        "rebalance";
+                }
+        }
+
+        if (assign_errstr) {
+                cfl_PyErr_Format(RD_KAFKA_RESP_ERR__FAIL, assign_errstr);
+		CallState_crash(cs);
+                CallState_resume(cs);
+                return;
+        }
+
 	/* Fallback: librdkafka needs the rebalance_cb to call assign()
 	 * to synchronize state, if the user did not do this from callback,
 	 * or there was no callback, or the callback failed, then we perform
 	 * that assign() call here instead. */
-	if (!self->u.Consumer.rebalance_assigned) {
-		if (err == RD_KAFKA_RESP_ERR__ASSIGN_PARTITIONS)
-			rd_kafka_assign(rk, c_parts);
-		else
-			rd_kafka_assign(rk, NULL);
+	if (!(self->u.Consumer.rebalance_assigned ||
+              self->u.Consumer.rebalance_incremental_assigned ||
+              self->u.Consumer.rebalance_incremental_unassigned)) {
+                if (!strcmp(rd_kafka_rebalance_protocol(rk), "COOPERATIVE")) {
+		        if (err == RD_KAFKA_RESP_ERR__ASSIGN_PARTITIONS)
+			        error = rd_kafka_incremental_assign(rk,
+                                                                    c_parts);
+                        else
+                                error = rd_kafka_incremental_unassign(rk,
+                                                                      c_parts);
+                } else {
+		        if (err == RD_KAFKA_RESP_ERR__ASSIGN_PARTITIONS)
+			        assign_err = rd_kafka_assign(rk, c_parts);
+		        else
+			        assign_err = rd_kafka_assign(rk, NULL);
+                }
 	}
 
+	if (assign_err) {
```

Comment:
Keep this inside the scope where it is used, including the variable defs, don't think this will ever be true if the fallback hasn't been used.

Reply:
good point